src/txn/txn_ckpt.c

```diff
@@ -642,6 +644,50 @@ __checkpoint_prepare(
 	WT_RET(__wt_config_gets(session, cfg, "use_timestamp", &cval));
 	use_timestamp = (cval.val != 0);
 
+#ifdef HAVE_TIMESTAMPS
+	/*
+	 * Set the checkpoint transaction's timestamp, if requested.
+	 *
+	 * We rely on having the global transaction data locked so the oldest
+	 * timestamp can't move past the stable timestamp.
+	 */
+	WT_ASSERT(session, !F_ISSET(txn,
+	    WT_TXN_HAS_TS_COMMIT | WT_TXN_HAS_TS_READ |
+	    WT_TXN_PUBLIC_TS_COMMIT | WT_TXN_PUBLIC_TS_READ));
+
```

Comment:
Hmm, on looking more closely at this, I see that we did it to avoid the race mentioned in the comment above.  This is going to need some more work...